libs/auth/src/java_card.ts

```diff
@@ -0,0 +1,462 @@
+            return;
+          }
+          default: {
+            throwIllegalValue(readerStatus);
+          }
+        }
+      },
+    });
+  }
+
+  async getCardStatus(): Promise<CardStatus> {
+    return Promise.resolve(this.cardStatus);
+  }
+
+  async checkPin(pin: string): Promise<CheckPinResponse> {
+    const cardVxAdminCert = await this.retrieveCert(
+      CARD_VX_ADMIN_CERT.OBJECT_ID
+    );
+    try {
+      // Verify that the card has a private key that corresponds to the public key in the card
+      // VxAdmin cert
+      await this.verifyCardPrivateKey(
+        CARD_VX_ADMIN_CERT.PRIVATE_KEY_SLOT,
+        cardVxAdminCert,
+        pin
+      );
+    } catch (error) {
+      if (
+        error instanceof ResponseApduError &&
+        isIncorrectPinResponseApdu(error.statusWord())
+      ) {
+        return {
+          response: 'incorrect',
+          numRemainingAttempts:
+            numRemainingAttemptsFromIncorrectPinResponseApdu(
+              error.statusWord()
+            ),
+        };
+      }
+      return { response: 'error' };
+    }
+    return { response: 'correct' };
+  }
+
+  program(): Promise<void> {
+    return Promise.resolve();
+  }
+
+  readData(): Promise<Buffer> {
+    return Promise.resolve(Buffer.from([]));
+  }
+
+  writeData(): Promise<void> {
+    return Promise.resolve();
+  }
+
+  clearData(): Promise<void> {
+    return Promise.resolve();
+  }
+
+  unprogram(): Promise<void> {
+    return Promise.resolve();
+  }
+
+  /**
+   * We wrap this.readUser() because it:
+   * 1. Intentionally throws errors if any verification fails
+   * 2. Can throw errors due to external actions like preemptively removing the card from the card
+   *    reader
+   */
+  private async safeReadUser(): Promise<Optional<User>> {
+    try {
+      return await this.readUser();
+    } catch {
+      return undefined;
+    }
+  }
+
+  /**
+   * Reads the user on the card, performing various forms of verification along the way. Throws an
+   * error if any verification fails.
+   */
+  private async readUser(): Promise<Optional<User>> {
+    await this.selectApplet();
+
+    // Verify that the card VotingWorks cert was signed by VotingWorks
+    const cardVxCert = await this.retrieveCert(CARD_VX_CERT.OBJECT_ID);
+    await openssl([
+      'verify',
+      '-CAfile',
+      this.vxCertAuthorityCertPath,
+      cardVxCert,
+    ]);
+
+    // Verify that the card VxAdmin cert was signed by VxAdmin
+    const cardVxAdminCert = await this.retrieveCert(
+      CARD_VX_ADMIN_CERT.OBJECT_ID
+    );
+    const vxAdminCertAuthorityCert = await this.retrieveCert(
+      VX_ADMIN_CERT_AUTHORITY_CERT.OBJECT_ID
+    );
+    await openssl([
+      'verify',
+      '-CAfile',
+      vxAdminCertAuthorityCert,
+      cardVxAdminCert,
+    ]);
+
+    // Verify that the VxAdmin cert authority cert was signed by VotingWorks
+    // TODO: Figure out how to sign the VxAdmin cert authority cert with the VotingWorks cert
+    // authority cert
```

Comment:
This TODO won't affect this code but is just something that I need to figure out in terms of certificate setup